File: Python codes/FFT_DIF_Algorithm.py
# -*- coding: utf-8 -*-
"""
Created on Sun Nov 19 10:20:51 2017

@author: Swaroop
"""

import numpy as np
import math

def fft_stage(N,stage,x):
   w=twiddle_factor(N/2**(stage-1)) 
   #print stage
   y=np.zeros(N,dtype=complex)
   for j in range(2**(stage-1)):
       
     for i in range(0,N/(2**stage),1):
        m=2*j*N/(2**(stage)) 
        #m=N/(2**(stage-1))
        #print(j)
        y[i+m]=x[i+m]+x[i+m+N/(2**stage)]
        if (2**stage) != N:
           y[i+m+N/(2**stage)]=(x[i+m]-x[i+m+N/(2**stage)])*w[i] 
        else:
           y[i+m+N/(2**stage)]=(x[i+m]-x[i+m+N/(2**stage)])
        
   return(y)    
def twiddle_factor(N):
    w=np.zeros(N,dtype=complex)
    for i in range(N):
        c=2*np.pi/N
        w[i]=np.round((np.cos(c*i)+1j*np.sin(c*i)),3)
    return w 

def swap(x,N):
    for i in range(1,N/2,2):
        c=x[i]
        x[i]=x[i-1+N/2]
        x[i-1+N/2]=c
    return x

    
N=8

no_stages=int(math.log(N,2))

x=([1,2,3,4,5,6,7,8])
z=np.zeros((no_stages,N),dtype=complex)

#stage=np.linspace(1,no_stages,1)
for i in range(no_stages):
    if i==0:
        z[i]=fft_stage(N,i+1,x)
    else:    
        
      z[i]=fft_stage(N,i+1,z[i-1])
      
    
k=swap(z[i],N)            

#X=z[no_stages]

for i in range(no_stages):
    
    print(z[i])
    
print(np.fft.fft(x))

print(k)   

File: C codes/FFT_DIF_Algorithm.c
#include <stdio.h>
#include <complex.h>
#include <math.h>

#define PI 3.14159265

void twiddle_factor(float complex *p,int n)
{
int i;
float c;
c=2*PI/n;
for(i=0;i<n;i++)
  p[i]=cos(c*i) + (sin(c*i))*I;
}


void fft_stage(int N,int stage,float complex *x,float complex *y)
{
  int k=N/pow(2,stage-1);
  float complex w[k];
  int i,j,m;

  twiddle_factor(w,k);

for(j=0;j<(int)pow(2,stage-1);j++)
{
    for(i=0;i<(int)(N/pow(2,stage));i++)
    {
        m=2*j*N/pow(2,stage);
            y[i+m]=x[i+m]+x[i+m+(int)(N/pow(2,stage))];
        if((int)pow(2,stage)!=N)
            y[i+m+(int)(N/pow(2,stage))]=(x[i+m]-x[i+m+(int)(N/pow(2,stage))])*w[i];
        else
            y[i+m+(int)(N/pow(2,stage))]=(x[i+m]-x[i+m+(int)(N/pow(2,stage))]);
    }
}
}


int main(){

int N=8,i,j,no_stages;
float complex z[3][N],Y[N],t[N],x[N];
no_stages=round(log(N)/log(2));
for(i=0;i<N;i++)
    x[i]=(i+1)+0*I;

for(i=0;i<no_stages;i++)
{

   if(i==0)
       fft_stage(N,i+1,x,Y);
     else
     {
       for(j=0;j<N;j++)
         t[j]=Y[j];
       fft_stage(N,i+1,t,Y);
       for(j=0;j<N;j++)
         z[i-1][j]=t[j];
     }
}
for(i=0;i<N;i++)
    printf("%.2f %+.2fI \n",crealf(z[no_stages-1][i]),cimagf(z[no_stages-1][i]));
return(0);

}
